apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,503 @@
+        """Private. Read both axes in its entirety, ensure that soma_joinid is retained."""
+        futures = (
+            self._threadpool.submit(
+                self._read_axis_dataframe,
+                "obs",
+                self.experiment.obs,
+                self._query["obs"],
+                column_names=column_names.get("obs", None),
+            ),
+            self._threadpool.submit(
+                self._read_axis_dataframe,
+                "var",
+                self.experiment.ms[self.measurement_name].var,
+                self._query["var"],
+                column_names=column_names.get("var", None),
+            ),
+        )
+        concurrent.futures.wait(futures)
+        obs_table, var_table = (f.result() for f in futures)
+        return obs_table, var_table
+
+    def read(
+        self,
+        X_name: str,
+        *,
+        use_position_indexing: bool = False,
+        column_names: Optional[AxisColumnNames] = None,
+        X_layers: Optional[List[str]] = None,
+    ) -> ExperimentAxisQueryReadArrowResult:
+        """
+        Read the _entire_ query result into (in-memory) Arrow Tables. This is a
+        low-level routine intended to be used by loaders for other in-core
+        formats, such as AnnData, which can be created from the resulting Tables.
+        """
+        if column_names is None:
+            column_names = AxisColumnNames(obs=None, var=None)
+        X_collection = self.experiment.ms[self.measurement_name].X
+        X_layers = [] if X_layers is None else X_layers
+        all_X_names = [X_name] + X_layers
+        all_X_arrays: Dict[str, SOMASparseNDArray] = {}
+        for _xname in all_X_names:
+            if not isinstance(_xname, str) or not _xname:
+                raise ValueError("X layer names must be specified as a string.")
+            if _xname not in X_collection:
+                raise ValueError("Unknown X layer name")
+            if X_collection[_xname].soma_type != "SOMASparseNDArray" or not isinstance(
+                X_collection[_xname], SOMASparseNDArray
+            ):
+                raise NotImplementedError("Dense array unsupported")
+            all_X_arrays[_xname] = cast(SOMASparseNDArray, X_collection[_xname])
+
+        obs_table, var_table = self._read_both_axes(column_names)
+
+        X_tables = {
+            # TODO: could also be done concurrently
+            _xname: all_X_arrays[_xname]
+            .read((self.obs_joinids(), self.var_joinids()))
+            .tables()
+            .concat()
+            for _xname in all_X_arrays
+        }
+        if use_position_indexing:
+            X_tables = self._rewrite_X_for_positional_indexing(X_tables)
+
+        X = X_tables.pop(X_name)
+        query_result: ExperimentAxisQueryReadArrowResult = {
+            "obs": obs_table,
+            "var": var_table,
+            "X": X,
+        }
+        if len(X_layers) > 0:
+            assert len(X_layers) == len(X_tables)
```

Comment:
Just to make sure: This is an internal sanity check, and external user code should never be able to trigger this (because if so, it should instead be a `raise SomeKindOfError`, since `assert`ions are assumed to be able to be stripped out with no impact on correctness).

Reply:
yes, sanity.  will remove as it is really unnecessary.